bls/signature.go

```diff
@@ -0,0 +1,237 @@
+}
+
+func (privateKey *PrivateKey) PublicKey() *PublicKey {
+	if privateKey.flowPrivateKey == nil {
+		return nil
+	}
+	return &PublicKey{flowPublicKey: privateKey.flowPrivateKey.PublicKey()}
+}
+
+func (privateKey *PrivateKey) ToString() string {
+	if privateKey.flowPrivateKey == nil {
+		return ""
+	}
+	return privateKey.flowPrivateKey.String()
+}
+
+func (privateKey *PrivateKey) FromString(privateKeyString string) (*PrivateKey, error) {
+	if privateKeyString == "" {
+		return nil, nil
+	}
+	// Chop off leading 0x, if exists. Otherwise, does nothing.
+	privateKeyStringCopy, _ := strings.CutPrefix(privateKeyString, "0x")
+	// Convert from hex string to byte slice.
+	privateKeyBytes, err := hex.DecodeString(privateKeyStringCopy)
+	if err != nil {
+		return nil, err
+	}
+	// Convert from byte slice to bls.PrivateKey.
+	privateKey.flowPrivateKey, err = flowCrypto.DecodePrivateKey(SigningAlgorithm, privateKeyBytes)
+	return privateKey, err
+}
+
+func (privateKey *PrivateKey) Eq(other *PrivateKey) bool {
+	if privateKey.flowPrivateKey == nil || other == nil {
+		return false
+	}
+	return privateKey.flowPrivateKey.Equals(other.flowPrivateKey)
+}
+
+//
+// TYPES: PublicKey
+//
+
+type PublicKey struct {
+	flowPublicKey flowCrypto.PublicKey
+}
+
+func (publicKey *PublicKey) Verify(signature *Signature, input []byte) (bool, error) {
+	if publicKey.flowPublicKey == nil {
+		return false, errors.New("bls.PublicKey is nil")
+	}
+	return publicKey.flowPublicKey.Verify(signature.flowSignature, input, HashingAlgorithm)
+}
+
+func (publicKey *PublicKey) ToBytes() []byte {
+	var publicKeyBytes []byte
+	if publicKey.flowPublicKey != nil {
+		publicKeyBytes = publicKey.flowPublicKey.Encode()
+	}
+	return publicKeyBytes
+}
+
+func (publicKey *PublicKey) FromBytes(publicKeyBytes []byte) (*PublicKey, error) {
+	if len(publicKeyBytes) == 0 {
+		return nil, nil
+	}
+	var err error
+	publicKey.flowPublicKey, err = flowCrypto.DecodePublicKey(SigningAlgorithm, publicKeyBytes)
+	return publicKey, err
+}
+
+func (publicKey *PublicKey) ToString() string {
+	if publicKey.flowPublicKey == nil {
+		return ""
+	}
+	return publicKey.flowPublicKey.String()
+}
+
+func (publicKey *PublicKey) FromString(publicKeyString string) (*PublicKey, error) {
+	if publicKeyString == "" {
+		return nil, nil
+	}
+	// Chop off leading 0x, if exists. Otherwise, does nothing.
+	publicKeyStringCopy, _ := strings.CutPrefix(publicKeyString, "0x")
+	// Convert from hex string to byte slice.
+	publicKeyBytes, err := hex.DecodeString(publicKeyStringCopy)
+	if err != nil {
+		return nil, err
+	}
+	// Convert from byte slice to bls.PublicKey.
+	publicKey.flowPublicKey, err = flowCrypto.DecodePublicKey(SigningAlgorithm, publicKeyBytes)
+	return publicKey, err
+}
+
+func (publicKey *PublicKey) Eq(other *PublicKey) bool {
+	if publicKey.flowPublicKey == nil || other == nil {
+		return false
+	}
+	return publicKey.flowPublicKey.Equals(other.flowPublicKey)
+}
+
+func (publicKey *PublicKey) Copy() *PublicKey {
```

Comment:
not sure about this copy function, can we add a test to make sure we're not copying any pointers under the hood?